src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -295,6 +314,73 @@ void ScheduleTask(DateTimeOffset utcNow)
             }
         }
 
+        internal readonly List<WeakReference<Stats>?> _allStats = new();
+        internal readonly Stats _accumulatedStats;
+
+        private (int, int) Sum()
+        {
+            int hits = 0, misses = 0;
+            lock (_allStats)
+            {
+                hits += _accumulatedStats.Hits;
+                misses += _accumulatedStats.Misses;
+                foreach (WeakReference<Stats>? wr in _allStats)
+                {
+                    if (wr is not null && wr.TryGetTarget(out Stats? stats))
+                    {
+                        hits += stats.Hits;
+                        misses += stats.Misses;
+                    }
+                }
+            }
+            return (hits, misses);
+        }
+
+        private ThreadLocal<Stats> _stats;
+
+        private void Hit()
+        {
+            GetStats().Hits++;
+        }
+
+        private void Miss()
+        {
+            GetStats().Misses++;
+        }
+
+        private Stats GetStats() => _stats!.Value!;
+
+        internal sealed class Stats
+        {
+            public int Hits;
+            public int Misses;
```

Comment:
Refer to earlier comment in https://github.com/dotnet/runtime/pull/66479#issuecomment-1072874447

> one limitation to note is the hit/miss values are int rather than long.

Reply:
If you change them to be long (to avoid overflowing) you will also need to use Interlocked.Write/Read, at least on 32-bit, assuming we care about avoiding torn values. (This will have a perf impact on 32-bit, but will also avoid overflowing.)

Reply:
Is there a perf concern if they were long? With all the hot path writes being thread-local data I wouldn't expect it to make much difference performance wise and a long running service with heavy cache usage might overflow the int.

Reply:
Functional correctness is more important then perf, so if long is needed to be functionally correct, so be it. But we might find it adds enough overhead we need to make it opt-in, if we care about 32-bit perf and care about avoiding the possibility of torn reads/writes with whatever thread is monitoring the data. Making them long would require Interlocked.Writes on 32-bit if we want to avoid tears, and that would amount to effectively the overhead of taking an uncontended lock on every get.

Reply:
oh I was being dense, I see what you mean about the torn read now. Can we give a shot and see what it does to perf? I think we could make at least a weak argument that 32 bit wasn't incorrect, it was simply a known design limitation. The (hopefully small) group of customers who saw wildly wrong hit ratios probably wouldn't be happy with us though.